src/safir/redis.py

```diff
@@ -0,0 +1,235 @@
+
+    def __init__(self, msg: str, key: str) -> None:
+        super().__init__(msg)
+        self.key = key
+
+    def to_slack(self) -> SlackMessage:
+        message = super().to_slack()
+        message.fields.append(SlackTextField(heading="Key", text=self.key))
+        return message
+
+
+#: Type variable for the type of object being stored.
+S = TypeVar("S", bound="BaseModel")
+
+
+class PydanticRedisStorage(Generic[S]):
+    """JSON-serialized encrypted storage in Redis.
+
+    Parameters
+    ----------
+    datatype
+        The class of object being stored (a Pydantic model).
+    redis
+        A Redis client configured to talk to the backend store.
+    key_prefix
+        A prefix to prepend to all Redis keys. Setting a prefix ensures that
+        all objects stored through this class share the same redis key
+        prefix. This is useful if multiple storage classes share the same
+        Redis database. The keys that you use to store objects (e.g. the
+        `store` method) and the key you get back from the `scan` method do
+        not include this prefix.
+    """
+
+    def __init__(
+        self, *, datatype: type[S], redis: redis.Redis, key_prefix: str = ""
+    ) -> None:
+        self._datatype = datatype
+        self._redis = redis
+        self._key_prefix = key_prefix
+
+    async def delete(self, key: str) -> bool:
+        """Delete a stored object.
+
+        Parameters
+        ----------
+        key
+            The key to delete.
+
+        Returns
+        -------
+        bool
+            `True` if the key was found and deleted, `False` otherwise.
+        """
+        count = await self._redis.delete(self._prefix_key(key))
+        return count > 0
+
+    async def delete_all(self, pattern: str) -> None:
+        """Delete all stored objects.
+
+        Parameters
+        ----------
+        pattern
+            Glob pattern matching the keys to purge, such as ``oidc:*``.
+        """
+        async for key in self._redis.scan_iter(self._prefix_key(pattern)):
+            await self._redis.delete(key)
+
+    async def get(self, key: str) -> S | None:
+        """Retrieve a stored object.
+
+        Parameters
+        ----------
+        key
+            The key for the object.
+
+        Returns
+        -------
+        Any or None
+            The deserialized object or `None` if no such object could be
+            found.
+
+        Raises
+        ------
+        DeserializeError
+            Raised if the stored object could not be decrypted or
+            deserialized.
+        """
+        full_key = self._prefix_key(key)
+        data = await self._redis.get(full_key)
+        if not data:
+            return None
+
+        try:
+            return self._deserialize(data)
+        except Exception as e:
+            error = f"{type(e).__name__}: {str(e)}"
+            msg = f"Cannot deserialize data for key {full_key}: {error}"
+            raise DeserializeError(msg, key=full_key) from e
```

Comment:
So here this would look like:
```suggestion
            error = f"{type(e).__name__}: {str(e)}"
            msg = f"Cannot deserialize data for key {full_key}"
            raise DeserializeError(msg, key=full_key, error=error) from e
```